### PR TITLE
[secretstore] migrate to version 4

### PR DESCRIPTION
**Pull request by ordian**
in preparation for https://github.com/paritytech/parity-common/pull/278

I'm deliberately not using `util/migration-rocksdb` here in order to avoid adding more dependencies on parity-ethereum to secret-store as it's about to be moved.
- [x] store version in a file

### CODE REVIEW

File: parity/db/rocksdb/migration_secretstore.rs [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+	}
+}
+
+impl From<IoError> for Error {
+	fn from(err: IoError) -> Self {
+		Error::Io(err)
+	}
+}
+
+// Moved "default" column to column 0 in preparation for a kvdb-rocksdb 0.3 migration.
```
Comment (dvdplm) [inline parity/db/rocksdb/migration_secretstore.rs]:
```suggestion
// Moves "default" column to column 0 in preparation for a kvdb-rocksdb 0.3 migration.
```

File: parity/db/rocksdb/migration_secretstore.rs [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+	const BATCH_SIZE: usize = 1024;
+	let new_dir = migration_dir(parent_dir);
+	let old_db = Database::open(&old_db_config, &db_dir(parent_dir))?;
+	let new_db = Database::open(&new_db_config, &new_dir)?;
+
+	// remove legacy version key
+	{
+		let mut batch = DBTransaction::with_capacity(1);
+		batch.delete(None, LEGACY_DB_META_KEY_VERSION);
+		old_db.write(batch)?;
```
Comment (dvdplm) [inline parity/db/rocksdb/migration_secretstore.rs]: Worth logging here imo (I don't think the error from `write` is very specific): `error!("db version deletion failed: {}")` or something?

File: parity/db/rocksdb/migration_secretstore.rs [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+	}
+
+	let mut batch = DBTransaction::with_capacity(BATCH_SIZE);
+	for (i, (key, value)) in old_db.iter(None).enumerate() {
+		batch.put(Some(0), &key, &value);
+		if i % BATCH_SIZE == 0 {
+			new_db.write(batch)?;
```
Comment (dvdplm) [inline parity/db/rocksdb/migration_secretstore.rs]: maybe some progress logging here e.g. `info!("Migrating secret-store db: {} keys written", BATCH_SIZE)`

File: secret-store/Cargo.toml [hunk 4]
```diff
@@ -17,6 +17,7 @@ ethcore-accounts = { path = "../accounts", optional = true}
 ethcore-call-contract = { path = "../ethcore/call-contract" }
 ethcore-sync = { path = "../ethcore/sync" }
 ethereum-types = "0.8.0"
+ethkey = { path = "../accounts/ethkey", optional = true }
```
Comment (dvdplm) [inline secret-store/Cargo.toml]: I had to add this too, to get `parity-ethereum` to build. A bit concerning CI didn't catch this.

**Comment by dvdplm** [comment]
Test failure looks legit.

File: parity/db/rocksdb/migration_secretstore.rs [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+use std::fmt::{Display, Error as FmtError, Formatter};
+use std::fs;
+use std::io::{Error as IoError, ErrorKind as IoErrorKind, Read as _, Write as _};
+use std::path::PathBuf;
+
+use kvdb::DBTransaction;
+use super::kvdb_rocksdb::{Database, DatabaseConfig};
+
+/// We used to store the version in the database (until version 4).
+const LEGACY_DB_META_KEY_VERSION: &'static [u8; 7] = b"version";
```
Comment (niklasad1) [inline parity/db/rocksdb/migration_secretstore.rs]:
```suggestion
const LEGACY_DB_META_KEY_VERSION: &[u8; 7] = b"version";
```

File: parity/db/rocksdb/migration_secretstore.rs [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+/// Current db version.
+const CURRENT_VERSION: u8 = 4;
+/// Database is assumed to be at the default version, when no version file is found.
+const DEFAULT_VERSION: u8 = 3;
+/// Version file name.
+const VERSION_FILE_NAME: &'static str = "db_version";
```
Comment (niklasad1) [inline parity/db/rocksdb/migration_secretstore.rs]:
```suggestion
const VERSION_FILE_NAME: &str = "db_version";
```

**Pull request merged**